API/ENH: add method='nearest' to Index.get_indexer/reindex and method to get_loc

shoyer opened the pull request:
Fixes #8845

Currently only an index method; I'm open to ideas for how to expose it more
directly -- maybe `df.loc_nearest`?.

In particular, I think this is usually a more useful way to do indexing for
floats than pandas's reindex based `.loc`, because floats are inherently
imprecise.

CC @jreback @jorisvandenbossche @immerrr @hugadams

shoyer commented:
Hmm. I think this has all the functionality we need, but it would also be good to reconcile the API design here with `asof`. I'll dig into that, but I'm also open to any proposals.

jorisvandenbossche commented:
- Should it return an array or an Index when given multiple values?
- For `asof`, I think this can completely replace the usage of `asof`, and is also compatible in signature. Only differences are:
  - `asof` would have a default of `left` in terms of `get_nearest`
  - it returns NaN if it cannot find anything. 

 I think we can leave `asof` as is, but add above remarks to the docs. Referring in the docstring of `asof` to `get_nearest` and indicating how you can replace it.

jreback commented:
I agree with @jorisvandenbossche this should just replace asof entirely
asod was essentially a simple version of this

jreback commented:
I think you should extract the indexer into another method maybe asof_indexer (and asof just calls that)
then we can support 'asof' joins! (I think there is an outstanding issue for this)

shoyer commented:
> Should it return an array or an Index when given multiple values?

I'm returning an array, based on how `get_indexer` works.

I'm currently modeling this function as a sort of combination of `get_loc` and `get_indexer`. Perhaps it would indeed make more sense to separate this into two interfaces -- `get_loc_nearest` (could handle duplicate indexes) and `get_indexer_nearest`.

jreback commented:
@shoyer yes this should be 2 methods ``get_indexer_nearest`` which is called by ``get_loc_nearest`` (whose impl is trivial), but this exposes the indexer which is important.

my 2c, is that we already have ``asof`` so this should be just an impl, e.g. ``get_loc_asof/get_indexer_asof``. No reason to have both

shoyer commented:
OK, latest commits include the split to 2 methods and handle missing values properly (that was not being done right before). The differences are:

1. `get_loc_nearest` will coerce input to the appropriate Timedelta/Timestamp/Period scalar like `get_loc`
2. `get_indexer_nearest` uses -1 to mark missing values that cannot be found whereas `get_loc_nearest` raises `KeyError`

I was able to make `Index.asof` be a thin wrapper around `get_loc_nearest`. So it should be able to handle non-datetimes now (for whatever that's worth).

The other use of asof is in the method `Series.asof` and the undocumented method `Index.asof_locs` (which appears to only be used or tested by `Series.asof`). This method is a little different and somewhat strange IMO -- it skips missing values in the Series. For now I've left it alone because the skipping of missing values makes it somewhat different, but is could definitely also use a cleanup.

hugadams commented:
Where do you find the time man!

shoyer commented:
One more design question: How should we handle partial string matching for `DatetimeIndex.get_loc_nearest`?

The current design casts the input to a Timestamp and finds the nearest label. This is a fine rule, but it's not consistent with `get_loc`, which will return all matching labels (even a slice, if possible). Arguably, `get_loc_nearest` should be equivalent to calling `get_loc` and then falling back to nearest lookup if the key is not found.

Here's an concrete example:
```
idx = pd.to_datetime(['2000-01-31', '2000-02-28'])
print idx.get_loc_nearest('2000-02', side='left')
print idx.get_loc_nearest('2000-02', side='nearest')
```

With the current implementation, each of these would print 0. This is consistent with the current behavior of `asof` (as verified by `test_asof_datetime_partial` in `test_index.py`), but not `get_loc`, which would print 1 for each of these.

shoyer commented:
OK, it looks like I need to make my own API decision :).

I've settled on using the result of `get_loc` for the result of `DatetimeIndex.get_loc_nearest`, if possible. This means that the "nearest" value to a partial string index like `'2000-01'` will be any values in January 2001, not the nearest value to `2000-01-01T00:00:00`.

This is also a small breaking API change for `DatetimeIndex.asof`. I've added documentation to that effect to what's new in the latest commit (along with more functionality/tests).

Any other comments? I think this is possibly ready to merge (once rebased).

jorisvandenbossche commented:
Would it be much trouble to not have this API change in `asof`? (possibly just leaving it as it is if it causes too much problems)

jorisvandenbossche commented on pandas/core/index.py:
```diff
@@ -1083,22 +1083,18 @@ def identical(self, other):
+        the passed label. Return NaN if not found.
 
-        return label
+        See also: get_loc_nearest
```
can you make this a `See also` section?

```
See also
--------
get_loc_nearest : small explanation what this is
```

then you get automatic linking in the html pages

jorisvandenbossche commented on pandas/core/index.py:
```diff
@@ -2180,6 +2182,129 @@ def slice_locs(self, start=None, end=None, step=None):
+
+    def get_loc_nearest(self, key, side='nearest'):
+        """
+        Compute the locations of the nearest index values.
```
State this as singular? `Compute location of nearest index value`

jorisvandenbossche commented:
A question: when you have a partial string, what is the 'nearest' value, so what will `DatetimeIndex.get_loc_nearest` return? (say if you have all days of the month in your index, what day is 'nearest' to the month)?

shoyer commented:
@jorisvandenbossche We definitely could include a work around to avoid the change for `asof` by overwriting the subclass `DatetimeIndex.asof` method.... but I'm not even entirely sure this should really be considered an API change rather than a bug fix (even though there *was* a test for it).

Note the `asof` docstring: "For a sorted index, return the most recent label up to and including the passed label." A string like '2000-02' arguably does include the timestamp '2000-02-28'.

Looking into the history a little more: this test appeared in PR #8246 by @TomAugspurger. There, it appears that the issue was something else entirely -- `asof` was returning time stamps not in the index at all (see #8245). It's not clear to me that anyone carefully considered exactly *how* partial datetime strings should be interpreted in this change.

shoyer commented:
@jorisvandenbossche The "nearest" value from `idx.get_loc_nearest('2000-01')` will be a slice object indexing all values in January 2001 if `idx.get_loc('2000-01')` is valid; otherwise, it will return the location of the nearest value to `2000-01-01T00:00:00`.

jreback commented:
I haven't really looked at this in depth - but

the string slicing should be analogous to partial string indexing with the same semantics

at first glance it appears that they are the same?

shoyer commented:
@jreback Yes, the nearest partial string indexing of a DatetimeIndex is exactly equivalent to partial partial string indexing if there is a match; otherwise, it casts the string label to a Timestamp and finds the nearest value.

jorisvandenbossche commented on pandas/tseries/index.py:
```diff
@@ -1256,6 +1256,19 @@ def get_loc(self, key):
             except (KeyError, ValueError):
                 raise KeyError(key)
 
+    def get_loc_nearest(self, key, side='nearest'):
```
These should also get a docstring (there are normally some decorators for adding a docstring)

jorisvandenbossche commented:
@shoyer on `asof`, OK then, given this is rather new behaviour as you pointed out, so lets regard that as a bug fix (maybe also specify it like that in the whatsnew docs)

Some other things:

- On the partial string indexing. The docstring of `get_loc_nearest` says it returns an int -> so this is not always correct then.
- For a DatetimeIndex, you only implemented `get_loc_nearest` and not `get_indexer_nearest`. Is this on purpose? If you now try using `get_indexer_nearest` with a string, you get rather unhelpful error messages.

I know you modelled the functions to the existing `get_loc` and `get_indexer`, but supposing these wouldn't exist, I would have the following remarks:

- Why have both? Why not just one function that accepts label or array of labels? (and returns int or array of ints)
- I find the name a bit confusing. `get_loc_nearest` returns the 'integer location by position'. But in the 'indexing terminology', `loc` is for the 'label-location' based indexer.
- The return of `-1` if a key is not found, seems a bit strange for me. As you would typically use the result to index your frame/series. But -1 has in that case a specific meaning, namely the last value. And that is not always what you want I think. (I know this is not that simple, as you cannot have NaN in integer, and you cannot have non-integer in the indexer, ..)

jreback commented:
@jorisvandenbossche 

I think this impl is fine. The signal of -1 is typical in the way the indexers work. These are by definition ``int64`` and -1 is a not-found value. It is up to the user to process these.

By definition these are always indexers, by position, always, full-stop. These only have a single semantic for the output results. ``loc`` here is not location based, but just location (its an older nomenclature and I can see it being somewhat confusing, but its how all of the Index classes work.)

@shoyer I would think this API simpler if you do this;

```
def get_indexer(self, target, method=None, limit=None):   
        """
        Compute indexer and mask for new index given the current index. The
        indexer should be then used as an input to ndarray.take to align the
        current data to the new index. The mask determines whether labels are
        found or not in the current index

        Parameters
        ----------
        target : Index
        method : {'pad', 'ffill', 'backfill', 'bfill'}
            pad / ffill: propagate LAST valid observation forward to next valid
            backfill / bfill: use NEXT valid observation to fill gap
```

and allow ``method='nearest'`` (and allow the ``side`` parameter to be passed as well).
(similarly for ``get_loc``). 

I think its just think its adding API noise otherwise. These conceptually do exactly the same thing. (you certainly could have an internal ``_get_indexer_nearest`` if that makes the impl clearer though).

shoyer commented:
@jreback This is a good point about incorporating this into the `get_indexer` method. In fact, I realize now that `get_indexer_nearest(side='left')` is actually just a slower version of `get_indexer(method='pad')`. So it looks like adding `method='nearest'` is probably the way to go here.

shoyer commented:
It appears that the existing `get_indexer` method only works for monotonic ascending indexes and keys. So I'll probably keep around some of this code as a fallback option for `pad`/`backfill` with decreasing indexes and/or unordered keys.

shoyer commented:
@jorisvandenbossche I agree, the current index API is a little strange -- some refactoring to provide a unified API and to untangle `core.indexers` would definitely be in order. See also: #7651.

I'm implementing only the low level methods now, but eventually we should figure out a higher level API for nearest neighbor lookups in pandas.

shoyer commented:
OK, I did a big refactor here and moved everything around to match @jreback's suggested organization (I also did a rebase since the old changes are almost all moved). This yielded the nice benefit of supporting non-monotonic targets for pad/backfill with get_indexer, so I put that down as an enhancement (not sure if pandas currently uses that directly or not; probably not).

I was disappointed by the lack of direct test coverage I encountered for the existing `get_loc` and `get_indexer` methods. I added some in but assuredly we could do more. In particular, the index methods currently allows integer values ( in nanoseconds) to pass through as valid values for the time/delta/period indexes. I know most of that is currently checked by `core.indexers` when used with `.loc`, etc., but that is poor separation of concerns -- that logic should all be on the index.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -205,6 +218,22 @@ Bug Fixes
+  .. ipython:: python
+
+    pd.to_datetime(['2000-01-31', '2000-02-28']).asof('2000-02')
+
```
I guess this was a bug then?

jreback commented on pandas/core/index.py:
```diff
@@ -1085,22 +1085,20 @@ def identical(self, other):
+        try:
+            loc = self.get_loc(label, method='pad')
+        except KeyError:
+            return np.nan
```
this should be ``_get_na_value(self.dtype)``

jreback commented on pandas/core/index.py:
```diff
@@ -3945,20 +4026,15 @@ def get_indexer(self, target, method=None, limit=None):
+            indexer = self_index._get_fill_indexer(target, method, limit)
+        elif method == 'nearest':
+            raise NotImplementedError("method='nearest' not implemented yet "
+                                      'for MultiIindex; see GitHub issue 9365')
```
MultiIndex

jreback commented on pandas/core/index.py:
```diff
@@ -3945,20 +4026,15 @@ def get_indexer(self, target, method=None, limit=None):
+            indexer = self_index._engine.get_indexer(target.values)
         else:
-            indexer = self_index._engine.get_indexer(target_index.values)
+            raise ValueError('unrecognized method: %s' % method)
```
maybe make a general check for the method (IIRC there was one somewhere), and do this first (and raise the ValueError), so the internal methods can be assured of a good method.

jreback replied:
use ``self._get_method(method)`` at the top of method and remove the else block

jreback commented on pandas/tests/test_index.py:
```diff
@@ -1874,6 +1954,49 @@ def test_numeric_compat(self):
                 self.assertRaises(TypeError, f)
 
+    def test_get_loc(self):
+        idx = pd.date_range('2000-01-01', periods=3)
```
maybe tests with tz-aware index (result should be the same w.r.t. to the tz)

jreback commented on pandas/tseries/index.py:
```diff
@@ -1636,9 +1639,6 @@ def indexer_at_time(self, time, asof=False):
         ----------
         time : datetime.time or string
-        tz : string or pytz.timezone or dateutil.tz.tzfile
-            Time zone for time. Corresponding timestamps would be converted to
```
was this just wrong?

shoyer replied:
Yes, I think so. It's describing a parameter that does not exist on the method.

jreback commented on pandas/tslib.pyx:
```diff
@@ -1154,8 +1154,10 @@ cdef convert_to_tsobject(object ts, object tz, object unit):
-    else:
+    elif getattr(ts, '_typ', None) == 'period':
         raise ValueError("Cannot convert Period to Timestamp unambiguously. Use to_timestamp")
+    else:
```
why is this here?

shoyer replied:
It's a long story. Ultimately the issue is that `SparseSeries.__getitem__` relies on the questionable assumption that non-scalar input like `slice(2)` will raise a `TypeError` when used with `Index.get_loc`. My refactor to `DatetimeIndex.get_loc` meant that these started raising `KeyError`, because I needed to swallow earlier `TypeError` that arise in `get_loc` with `method='nearest'` (e.g., when attempting to use the original label when it's a string, not a Timestamp).

I'll add a test to `DatetimeIndex.get_loc` to verify that it raises TypeError for invalid input. `SparseSeries.__getitem__` could also use a refactor to rely less on exceptions for data flow but I'd rather not touch it if possible.

jreback replied:
ok why don't u create an issue (and point to some code) so that this can be fixed later (sparse certainly needs work)

shoyer replied:
done: #9467.

jreback commented on pandas/tests/test_index.py:
```diff
@@ -418,7 +418,7 @@ def test_is_(self):
 
```
need a test for passing invalid ``method`` (put in Base so its tested for all Index classes). Though since you override the checking, might be necessary to have some individual tests for this.

jreback commented on pandas/core/index.py:
```diff
@@ -1398,15 +1396,34 @@ def sym_diff(self, other, result_name=None):
+            pad / ffill: find the PREVIOUS index value if no exact match.
+            backfill / bfill: use NEXT index value if no exact match
+            nearest: use the NEAREST index value if no exact match. Tied
+            distances are broken by preferring the larger index value.
```
I suppose that one would maybe want to specify how things are tied (e.g. a side argument), but for a future PR

jreback commented on pandas/core/index.py:
```diff
@@ -1398,15 +1396,34 @@ def sym_diff(self, other, result_name=None):
+        key : label
+        method : {None, 'pad'/'ffill', 'backfill'/'bfill', 'nearest'}
+            default: exact matches only.
+            pad / ffill: find the PREVIOUS index value if no exact match.
```
if you want to get fancy, you can put the method docs in a module level variable (and use ``@Appender``), just a thought

jreback commented on pandas/core/index.py:
```diff
@@ -1494,7 +1512,10 @@ def get_indexer(self, target, method=None, limit=None):
+            positions matches the corresponding target values. Missing values
+            in the target are marked by -1.
         """
         method = self._get_method(method)
```
right, this should raise ``ValueError`` if its not valid (I made a comment below about this). So method should be good after here.

jreback commented on pandas/core/index.py:
```diff
@@ -1527,6 +1544,64 @@ def get_indexer(self, target, method=None, limit=None):
         return com._ensure_platform_int(indexer)
 
+    def _get_fill_indexer(self, target, method, limit=None):
+        if self.is_monotonic and target.is_monotonic:
```
dont' these have to be both monotonic_increasing (or both monotonic_decreasing)? e.g. if one is increasing and one decreasing is that an error?

shoyer replied:
`is_monotonic` is currently an alias for `is_monotonic_increasing`. I'll switch to the later to be more explicit.

jreback commented on pandas/tseries/period.py:
```diff
@@ -975,7 +975,13 @@ def get_value(self, series, key):
+        if hasattr(target, 'freq') and target.freq != self.freq:
+            raise ValueError('target and index have different freq: '
+                             '(%s, %s)' % (target.freq, self.freq))
+        return Index.get_indexer(self, target, method, limit)
```
use super here?

jreback commented:
@shoyer looks much better. consolidates the functionaility pretty nicely. Just some minor comments.

shoyer commented:
It looks like we can very easily expose this directly through `DataFrame.reindex` with `method='nearest'` (most of the current test failures are because `reindex` with `method='pad' no longer fails for non-monotonic targets). We may eventually want some sort of "nearest neighbor" indexer, but `reindex` is a useable API, so I will see if I can get that working at least.

jreback commented:
as a side issue

a time like join is essentially nearest_left (eg nearest but only looking left)
so maybe want to add nearest_left/nearest_right 
the can easily do : http://stackoverflow.com/questions/12322289/kdb-like-asof-join-for-timeseries-data-in-pandas

shoyer commented:
@jreback  OK, I think I address most of your suggestions in my latest patchset. I'm also presenting the public API using `reindex` instead of the low-level `get_indexer`.

We should also add this to the [docs section on reindexing](http://pandas-docs.github.io/pandas-docs-travis/basics.html#filling-while-reindexing), at least once fillna with `method='nearest'` also works (made a new issue for that, #9471)

jreback commented:
yeh I would update the docs once #9471 is addressed.

jreback commented on pandas/core/common.py:
```diff
@@ -2646,7 +2646,7 @@ def _astype_nansafe(arr, dtype, copy=True):
 
 
-def _clean_fill_method(method):
+def _clean_fill_method(method, allow_nearest=False):
```
this seems kind of hacky, why don't we just always allow nearest (I know its not quite supported by fillna). Or will you just fix this then?

shoyer replied:
yes, when nearest is valid for fillna we should remove this method.

I was considering adding nearest for fillna in this PR but I'd rather save it for another just to minimize the scope here.

shoyer commented:
Other thoughts for this issue? `fillna` will come in another PR.

jreback commented on pandas/core/generic.py:
```diff
@@ -1744,9 +1744,8 @@ def _reindex_axes(self, axes, level, limit, method, fill_value, copy):
-                fill_value=fill_value, limit=limit, copy=copy,
-                allow_dups=False)
+                {axis: [new_index, indexer]},
+                fill_value=fill_value, copy=copy, allow_dups=False)
```
doesn't this need method passed thru?

shoyer replied:
You'll note that `_reindex_with_indexers` doesn't actually use `method` or `limit` arguments. So I removed them from the function signature below.

jreback replied:
yep

jreback commented on pandas/core/generic.py:
```diff
@@ -1802,15 +1801,14 @@ def reindex_axis(self, labels, axis=0, method=None, level=None, copy=True,
         return self._reindex_with_indexers(
-            {axis: [new_index, indexer]}, method=method, fill_value=fill_value,
-            limit=limit, copy=copy)
+            {axis: [new_index, indexer]}, fill_value=fill_value, copy=copy)
```
oh, right, ``method`` is extraneous here (above too?)

jreback commented on pandas/core/index.py:
```diff
@@ -3945,20 +4014,13 @@ def get_indexer(self, target, method=None, limit=None):
+        if method == 'pad' or method == 'backfill':
+            indexer = self_index._get_fill_indexer(target, method, limit)
+        elif method == 'nearest':
+            raise NotImplementedError("method='nearest' not implemented yet "
```
I think you have a test for this, right?

shoyer replied:
Yes, and for `MultiIndex.get_loc`, too

jreback commented:
minor comments, lgtm otherwise.

jorisvandenbossche commented on pandas/core/index.py:
```diff
@@ -1473,19 +1490,20 @@ def get_indexer(self, target, method=None, limit=None):
+            pad / ffill: find the PREVIOUS index value if no exact match.
+            backfill / bfill: use NEXT index value if no exact match
+            nearest: use the NEAREST index value if no exact match. Tied
+            distances are broken by preferring the larger index value.
```
For the html docs, you have to make bullet points of these (use `-` or `*` before each line) to have this render properly. 
The same for the other docstrings.

jorisvandenbossche commented:
docstring of reindex should still be updated

shoyer commented:
Yep, the docstrings need a bit of work. I'll do that next week -- going camping and will be offline for the next three days.

shoyer committed:
API/ENH: add method='nearest' for Index.get_indexer and reindex

Fixes GH8845

shoyer commented:
OK, I updated docstrings for reindex, reindex_like, get_indexer and get_loc, and they appear to generate well-formatted API docs. Merge ready, I think?

jreback commented on pandas/tests/test_series.py:
```diff
@@ -5878,6 +5879,16 @@ def test_reindex_pad(self):
+        s = Series(np.arange(10))
+        target = [0.1, 0.9, 1.5, 2.0]
+        actual = s.reindex(target, method='nearest')
+        expected = Series(np.around(target).astype(int), target)
```
astype these to int64 otherwiset hey will fail on windows.

jreback commented on pandas/tests/test_series.py:
```diff
@@ -5878,6 +5879,16 @@ def test_reindex_pad(self):
         assert_series_equal(result, expected)
 
+    def test_reindex_nearest(self):
+        s = Series(np.arange(10))
```
specify dtype='int64'

jreback commented:
lgtm

- minor dtype chaneges needed in some tests
- haven't checked, but make basics.rst docs need updating (where reindex is, to maybe have an example for nearest)

shoyer commented:
@jreback Fixed the dtypes.

@jorisvandenbossche @jreback I was planning on updating the basics.rst docs later (once "nearest" works for fillna), but it was simple enough to do it now. Let me know what you think. In particular, I removed some musing from Wes about the possibility of more sophisticated interpolation logic.

jreback commented:
lgtm

shoyer commented:
@jorisvandenbossche do you want to take another look? otherwise I will merge...

shoyer commented:
OK, I'm going to merge this. Doc issues can be dealt with in a followup PR if necessary.

jreback commented:
nice work @shoyer !

mjroth commented:
Is the intent for this to return the nearest *valid* observation?  
i.e. for the example given in the whats new 0.16, 
<http://pandas-docs.github.io/pandas-docs-travis/whatsnew.html>
if we throw in a NaN, it does not search for the next valid observation.

```python
In [1]: import pandas as pd
In [2]: import numpy as np
In [3]: df = pd.DataFrame({'x': range(5)})
In [4]: df.x[0] = np.nan
In [5]: df.reindex([0.2, 1.8, 3.5], method='nearest')
Out[5]: 
      x
0.2 NaN
1.8   2
3.5   4
```
I'd expect x=1 at 0.2, given the Docs.

shoyer commented:
@mjroth the intent is for it to return values at the nearest index label, similarly to the existing behavior for forward and backward filling reindexing methods. If you don't want to include NA values, you'll need to do a dropna first.
